src/turbostat.c

```diff
@@ -0,0 +1,1611 @@
+		delta->mperf = new->mperf - old->mperf;
+	} else {
+		if (!aperf_mperf_unstable) {
+			WARNING("turbostat plugin: APERF or MPERF went "
+				"backwards. Frequency results do not cover "
+				"the entire interval. Fix this by running "
+				"Linux-2.6.30 or later.");
+
+			aperf_mperf_unstable = 1;
+		}
+	}
+
+	/*
+	 * As counter collection is not atomic,
+	 * it is possible for mperf's non-halted cycles + idle states
+	 * to exceed TSC's all cycles: show c1 = 0% in that case.
+	 */
+	if ((delta->mperf + core_delta->c3 + core_delta->c6 + core_delta->c7) > delta->tsc)
+		delta->c1 = 0;
+	else {
+		/* normal case, derive c1 */
+		delta->c1 = delta->tsc - delta->mperf - core_delta->c3
+			- core_delta->c6 - core_delta->c7;
+	}
+
+	if (delta->mperf == 0) {
+		WARNING("turbostat plugin: cpu%d MPERF 0!", old->cpu_id);
+		delta->mperf = 1;	/* divide by 0 protection */
+	}
+
+	if (do_smi)
+		delta->smi_count = new->smi_count - old->smi_count;
+
+	return 0;
+}
+
+/**********************************
+ * Submitting the results (1 CPU) *
+ **********************************/
+
+/*
+ * Submit one gauge value
+ */
+static void
+turbostat_submit (const char *plugin_instance,
+	const char *type, const char *type_instance,
+	gauge_t value)
+{
+	value_list_t vl = VALUE_LIST_INIT;
+	value_t v;
+
+	v.gauge = value;
+	vl.values = &v;
+	vl.values_len = 1;
+	sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+	sstrncpy (vl.plugin, PLUGIN_NAME, sizeof (vl.plugin));
+	if (plugin_instance != NULL)
+		sstrncpy (vl.plugin_instance, plugin_instance, sizeof (vl.plugin_instance));
+	sstrncpy (vl.type, type, sizeof (vl.type));
+	if (type_instance != NULL)
+		sstrncpy (vl.type_instance, type_instance, sizeof (vl.type_instance));
+
+	plugin_dispatch_values (&vl);
+}
+
+/*
+ * Submit every data for a single CPU
+ *
+ * Core data is shared for all threads in one core: submitted only for the first thread
+ * Package data is shared for all core in one package: submitted only for the first thread of the first core
+ */
+static int
+submit_counters(struct thread_data *t, struct core_data *c, struct pkg_data *p)
+{
+	char name[DATA_MAX_NAME_LEN];
+	double interval_float;
+
+	interval_float = CDTIME_T_TO_DOUBLE(time_delta);
+
+	ssnprintf(name, sizeof(name), "cpu%02d", t->cpu_id);
+
+	if (!aperf_mperf_unstable)
+		turbostat_submit(name, "percent", "c0", 100.0 * t->mperf/t->tsc);
+	if (!aperf_mperf_unstable)
+		turbostat_submit(name, "percent", "c1", 100.0 * t->c1/t->tsc);
+
+	turbostat_submit(name, "frequency", "average", 1.0 / 1000000 * t->aperf / interval_float);
+
+	if ((!aperf_mperf_unstable) || (!(t->aperf > t->tsc || t->mperf > t->tsc)))
+		turbostat_submit(name, "frequency", "busy", 1.0 * t->tsc / 1000000 * t->aperf / t->mperf / interval_float);
+
+	/* Sanity check (should stay stable) */
+	turbostat_submit(name, "gauge", "TSC", 1.0 * t->tsc / 1000000 / interval_float);
+
+	/* SMI */
+	if (do_smi)
+		turbostat_submit(name, "current", NULL, t->smi_count);
```

Comment:
Speaking of cleaning, this probably needs to be changed: the SMI count has nothing to do with currents (I just used it because I didn't know what type to use): what do you recon?